internal/controller/gateway.go

```diff
@@ -45,35 +44,36 @@ type gatewayReconciler struct {
 	gatewayClassControllerName gatewayapi_v1alpha2.GatewayController
 }
 
-// NewGatewayController creates the gateway controller from mgr. The controller will be pre-configured
+// RegisterGatewayController creates the gateway controller from mgr. The controller will be pre-configured
 // to watch for Gateway objects across all namespaces and reconcile those that match class.
-func NewGatewayController(
+func RegisterGatewayController(
+	log logrus.FieldLogger,
 	mgr manager.Manager,
 	eventHandler cache.ResourceEventHandler,
 	statusUpdater k8s.StatusUpdater,
-	log logrus.FieldLogger,
 	gatewayClassControllerName string,
-	isLeader <-chan struct{},
-) (controller.Controller, error) {
+) error {
 	r := &gatewayReconciler{
-		ctx:                        context.Background(),
+		log:                        log,
 		client:                     mgr.GetClient(),
 		eventHandler:               eventHandler,
 		statusUpdater:              statusUpdater,
-		log:                        log,
 		gatewayClassControllerName: gatewayapi_v1alpha2.GatewayController(gatewayClassControllerName),
 	}
-	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
+	c, err := controller.NewUnmanaged("gateway-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if err := mgr.Add(&noLeaderElectionController{c}); err != nil {
+		return err
 	}
 
 	if err := c.Watch(
```

Comment:
also note that these sort of watches *should* work with dynamic updates to contour configuration since we pass in a predicate func that has access to the reconciler's data, we should be able to update that data with correct concurrency/locking mechanisms (could be other solutions, this first springs to mind, theoretically should be cheap since we shouldnt get too many config crd updates)

Reply:
a point in favor of doubling down on controller-runtime managed controllers/runnables, in case that was a concern @projectcontour/maintainers